File: java/Samples/Biometrics/CS/SimpleIrisesSampleCS/SegmentIris.cs
using System;
using System.Globalization;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Images;
using System.Linq;

namespace Neurotec.Samples
{
	public partial class SegmentIris : UserControl
	{
		#region Public constructor

		public SegmentIris()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private NBiometricClient _biometricClient;
		private NIris _iris;
		private NIris _segmentedIris;

		#endregion

		#region Public properties

		public NBiometricClient BiometricClient
		{
			get { return _biometricClient; }
			set { _biometricClient = value; }
		}

		#endregion

		#region Private methods

		private void OnSegmentCompleted(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnSegmentCompleted), r);
			}
			else
			{
				NBiometricTask task = _biometricClient.EndPerformTask(r);
				NBiometricStatus status = task.Status;
				if (status == NBiometricStatus.Ok)
				{
					NEAttributes attributes = _iris.Objects.First();
					// Segmented iris
					_segmentedIris = attributes.Child as NIris;
					irisView2.Iris = _segmentedIris;

					tbQuality.Text = attributes.Quality.ToString(CultureInfo.InvariantCulture);
					tbGrayLevelSpread.Text = attributes.GrayScaleUtilisation.ToString(CultureInfo.InvariantCulture);
					tbPupilToIrisRatio.Text = attributes.PupilToIrisRatio.ToString(CultureInfo.InvariantCulture);
					tbUsableIrisArea.Text = attributes.UsableIrisArea.ToString(CultureInfo.InvariantCulture);
					tbIrisScleraContrast.Text = attributes.IrisScleraContrast.ToString(CultureInfo.InvariantCulture);
					tbIrisPupilContrast.Text = attributes.IrisPupilContrast.ToString(CultureInfo.InvariantCulture);
					tbIrisPupilConcentricity.Text = attributes.IrisPupilConcentricity.ToString(CultureInfo.InvariantCulture);
					tbPupilBoundaryCircularity.Text = attributes.PupilBoundaryCircularity.ToString(CultureInfo.InvariantCulture);
					tbMarginAdequacy.Text = attributes.MarginAdequacy.ToString(CultureInfo.InvariantCulture);
					tbSharpness.Text = attributes.Sharpness.ToString(CultureInfo.InvariantCulture);
					tbInterlace.Text = attributes.Interlace.ToString(CultureInfo.InvariantCulture);

					btnSegmentIris.Enabled = false;
					btnSaveImage.Enabled = true;
				}
				else
				{
					MessageBox.Show(string.Format("Segmentation failed: {0}.", status), Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
				}
				btnOpenImage.Enabled = true;
			}
		}

		private void ResetControls()
		{
			irisView1.Iris = null;
			irisView2.Iris = null;

			tbQuality.Text = string.Empty;
			tbGrayLevelSpread.Text = string.Empty;
			tbPupilToIrisRatio.Text = string.Empty;
			tbUsableIrisArea.Text = string.Empty;
			tbIrisScleraContrast.Text = string.Empty;
			tbIrisPupilContrast.Text = string.Empty;
			tbIrisPupilConcentricity.Text = string.Empty;
			tbPupilBoundaryCircularity.Text = string.Empty;
			tbMarginAdequacy.Text = string.Empty;
			tbSharpness.Text = string.Empty;
			tbInterlace.Text = string.Empty;

			btnSaveImage.Enabled = false;
			btnSegmentIris.Enabled = false;
		}

		#endregion

		#region Private form events

		private void BtnOpenImageClick(object sender, EventArgs e)
		{
			ResetControls();

			if (openFileDialog.ShowDialog() == DialogResult.OK)
			{
				try
				{
					_iris = new NIris { Image = NImage.FromFile(openFileDialog.FileName) };
					irisView1.Iris = _iris;
					btnSegmentIris.Enabled = true;
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		private void BtnSegmentIrisClick(object sender, EventArgs e)
		{
			if (_iris != null)
			{
				_iris.ImageType = NEImageType.CroppedAndMasked;
				var subject = new NSubject();
				subject.Irises.Add(_iris);

				NBiometricTask segmentTask = _biometricClient.CreateTask(NBiometricOperations.Segment, subject);
				_biometricClient.BeginPerformTask(segmentTask, OnSegmentCompleted, null);
				btnSegmentIris.Enabled = false;
				btnOpenImage.Enabled = false;
			}
		}

		private void BtnSaveImageClick(object sender, EventArgs e)
		{
			if (_segmentedIris != null)
			{
				saveFileDialog.Filter = NImages.GetSaveFileFilterString();
				if (saveFileDialog.ShowDialog() == DialogResult.OK)
				{
					try
					{
						_segmentedIris.Image.Save(saveFileDialog.FileName);
					}
					catch (Exception ex)
					{
						Utils.ShowException(ex);
					}
				}
			}
		}

		private void SegmentIrisLoad(object sender, EventArgs e)
		{
			if (!DesignMode)
			{
				btnSaveImage.Enabled = false;
				btnSegmentIris.Enabled = false;
				saveFileDialog.Filter = NImages.GetSaveFileFilterString();
				openFileDialog.Filter = NImages.GetOpenFileFilterString(true, false);
			}
		}

		#endregion
	}
}

File: java/Samples/Biometrics/VB/SimpleIrisesSampleVB/SegmentIris.vb
Imports System
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Images
Imports System.Linq

Partial Public Class SegmentIris
	Inherits UserControl
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

#End Region

#Region "Private fields"

	Private _biometricClient As NBiometricClient
	Private _iris As NIris
	Private _segmentedIris As NIris

#End Region

#Region "Public properties"

	Public Property BiometricClient() As NBiometricClient
		Get
			Return _biometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			_biometricClient = value
		End Set
	End Property

#End Region

#Region "Private methods"

	Private Sub OnSegmentCompleted(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnSegmentCompleted), r)
		Else
			Dim task As NBiometricTask = _biometricClient.EndPerformTask(r)
			Dim status As NBiometricStatus = task.Status
			If status = NBiometricStatus.Ok Then
				Dim attributes As NEAttributes = _iris.Objects.First()
				' Segmented iris
				_segmentedIris = TryCast(attributes.Child, NIris)
				irisView2.Iris = _segmentedIris

				tbQuality.Text = attributes.Quality.ToString()
				tbGrayLevelSpread.Text = attributes.GrayScaleUtilisation.ToString()
				tbPupilToIrisRatio.Text = attributes.PupilToIrisRatio.ToString()
				tbUsableIrisArea.Text = attributes.UsableIrisArea.ToString()
				tbIrisScleraContrast.Text = attributes.IrisScleraContrast.ToString()
				tbIrisPupilContrast.Text = attributes.IrisPupilContrast.ToString()
				tbIrisPupilConcentricity.Text = attributes.IrisPupilConcentricity.ToString()
				tbPupilBoundaryCircularity.Text = attributes.PupilBoundaryCircularity.ToString()
				tbMarginAdequacy.Text = attributes.MarginAdequacy.ToString()
				tbSharpness.Text = attributes.Sharpness.ToString()
				tbInterlace.Text = attributes.Interlace.ToString()

				btnSegmentIris.Enabled = False
				btnSaveImage.Enabled = True
			Else
				MessageBox.Show(String.Format("Segmentation failed: {0}.", status), Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
			End If
			btnOpenImage.Enabled = True
		End If
	End Sub

	Private Sub ResetControls()
		irisView1.Iris = Nothing
		irisView2.Iris = Nothing

		tbQuality.Text = String.Empty
		tbGrayLevelSpread.Text = String.Empty
		tbPupilToIrisRatio.Text = String.Empty
		tbUsableIrisArea.Text = String.Empty
		tbIrisScleraContrast.Text = String.Empty
		tbIrisPupilContrast.Text = String.Empty
		tbIrisPupilConcentricity.Text = String.Empty
		tbPupilBoundaryCircularity.Text = String.Empty
		tbMarginAdequacy.Text = String.Empty
		tbSharpness.Text = String.Empty
		tbInterlace.Text = String.Empty

		btnSaveImage.Enabled = False
		btnSegmentIris.Enabled = False
	End Sub

#End Region

#Region "Private form events"

	Private Sub BtnOpenImageClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnOpenImage.Click
		ResetControls()

		If openFileDialog.ShowDialog() = DialogResult.OK Then
			Try
				_iris = New NIris With {.Image = NImage.FromFile(openFileDialog.FileName)}
				irisView1.Iris = _iris
				btnSegmentIris.Enabled = True
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

	Private Sub BtnSegmentIrisClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSegmentIris.Click
		If _iris IsNot Nothing Then
			_iris.ImageType = NEImageType.CroppedAndMasked
			Dim subject As New NSubject()
			subject.Irises.Add(_iris)

			Dim segmentTask As NBiometricTask = _biometricClient.CreateTask(NBiometricOperations.Segment, subject)
			_biometricClient.BeginPerformTask(segmentTask, AddressOf OnSegmentCompleted, Nothing)
			btnSegmentIris.Enabled = False
			btnOpenImage.Enabled = False
		End If
	End Sub

	Private Sub BtnSaveImageClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSaveImage.Click
		If _segmentedIris IsNot Nothing Then
			saveFileDialog.Filter = NImages.GetSaveFileFilterString()
			If saveFileDialog.ShowDialog() = DialogResult.OK Then
				Try
					_segmentedIris.Image.Save(saveFileDialog.FileName)
				Catch ex As Exception
					Utils.ShowException(ex)
				End Try
			End If
		End If
	End Sub

	Private Sub SegmentIrisLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		If (Not DesignMode) Then
			btnSaveImage.Enabled = False
			btnSegmentIris.Enabled = False
			saveFileDialog.Filter = NImages.GetSaveFileFilterString()
			openFileDialog.Filter = NImages.GetOpenFileFilterString(True, False)
		End If
	End Sub

#End Region
End Class
